File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: TreeNode) -> int:
        self.count = 0
        def dfs(node):
            if not node:
                return 0 , 0
            lval , lcount = dfs(node.left)
            rval , rcount = dfs(node.right)

            nodeVal = lval + rval + node.val
            nodeCount = lcount + rcount + 1

            if nodeVal // nodeCount == node.val: self.count += 1
            return nodeVal , nodeCount
        dfs(root)
        return self.count

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
typedef struct{
    int sum;
    int count;
}SubResult;

SubResult dfs(struct TreeNode* node , int* resultCount){
    SubResult res = {0 , 0};
    if(node == NULL){
        return res;
    }

    SubResult leftres = dfs(node->left , resultCount);
    SubResult rightres = dfs(node->right , resultCount);

    res.sum = leftres.sum + rightres.sum + node->val;
    res.count = leftres.count + rightres.count + 1;

    if(res.sum / res.count == node->val){
        (*resultCount)++;
    }

    return res;
}

int averageOfSubtree(struct TreeNode* root) {
    int resultCount = 0;
    dfs(root , &resultCount);
    return resultCount;
}